runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -365,7 +366,9 @@ end
 ---@return table Table with map of line number to list of diagnostics.
 ---              Structured: { [1] = {...}, [5] = {.... } }
 ---@private
+---@deprecated
 function M.get_line_diagnostics(bufnr, line_nr, opts, client_id)
```

Comment:
This function is the only way to get diagnostics as defined in LSP instead of vim diagnostics, and there are no public utilities to convert between them. Thoughts on exposing a way to convert between the different diagnostic formats?

Reply:
Is there a use case? This function has been private/undocumented since `vim.diagnostic` was released so nothing is changing here.

In general I think that if users need to interface with “diagnostics” they should do so through `vim.diagnostic`. `vim.lsp.diagnostic` is only for dealing with “LSP diagnostics”, which is a separate and narrowly scoped thing, and only used by the LSP client implementation.

Exposing a public interface for both `vim.diagnostic` *and* LSP diagnostics seems like a leaky abstraction. But without an actual use case to consider it’s hard to know for sure.

Reply:
It's usage is mostly relevant for plugins that deal with code actions. I personally [use it in my config](https://github.com/MariaSolOs/dotfiles/blob/40da0d80859b27b33b7d252677c448b74c813d65/.config/nvim/lua/lightbulb.lua#L45) to have a "VS Code-like lightbulb". The same is done by [`nvim-lightbulb`](https://github.com/kosayoda/nvim-lightbulb/blob/8f00b89dd1b1dbde16872bee5fbcee2e58c9b8e9/lua/nvim-lightbulb/init.lua#L260) and [`lspsaga`](https://github.com/nvimdev/lspsaga.nvim/blob/28051f1be64c85aaeb9272de072cd9a13dfa5150/lua/lspsaga/codeaction/lightbulb.lua#L59).

A GitHub search also shows that [`flutter-tools`](https://github.com/akinsho/flutter-tools.nvim/blob/7350750d46fbeb4d2bb4878157b658d435935299/lua/flutter-tools/lsp/code_actions.lua#L13) and [`navigator.nvim`](https://github.com/ray-x/navigator.lua/blob/4d6b1b804f0a0d0aee1de9d38b00bc0d3a2ee4e5/lua/navigator/codeAction.lua#L183) use it as well.

Reply:
But why aren't you using the `vim.diagnostic` API for that (honest question)? That should give you support for _any_ diagnostic, for free?

Reply:
@clason because those diagnostics are used in the code action request, and the language servers expect an LSP diagnostic, not vim ones.

Reply:
So the conversion to vim.diagnostic is lossy in this case?

Reply:
@clason Not sure what you mean exactly. What I’m trying to say here is that sometimes one might need diagnostics in their LSP format, yet all the functions that provided those got deprecated and `vim.diagnostic.get` is mentioned as the alternative. However that function returns vim diagnostics, and there’s no public API to convert between LSP and vim diagnostics.

Reply:
(Don't get me wrong: I think that this is a valid use case; I just want to explore options how to best enable it.)

Reply:
> there’s no public API to convert between LSP and vim diagnostics

What I'm kinda getting at is asking whether you _need_ an API to convert this, or whether it's feasible to do this by hand. Or whether you can completely encapsulate this so that no API call is needed.

Reply:
I think part of the usages are historic. I use it in nvim-jdtls too because it was written before vim.diagnostic was a thing.

Some usages can likely be replaced by using `code_action` with the `filter` / `apply` options.

For other use-cases I think we have several options:

- Attach the full lsp.Diagnostic object into the user_data of the vim.diagnostic entries. Avoids conversions but increases memory footprint
- Expose a function to convert between lsp.Diagnostic and vim.Diagnostic
- The `vim.pos` / `Position` story, and conversion from lsp.Position/vim.Position. Then plugin authors still have to convert from vim.Diagnostic to lsp.Diagnostic but at least the position conversion could be easier.
- Do nothing, let them re-implement it.

Reply:
>Attach the full lsp.Diagnostic object into the user_data of the vim.diagnostic entries. Avoids conversions but increases memory footprint

What if we included just the original `range`? That seems like the most difficult field to get back post-conversion, and is also likely the most useful for plugins.

Reply:
> What if we included just the original range? That seems like the most difficult field to get back post-conversion, and is also likely the most useful for plugins.

Good idea :+1: 

> I think this is my least favorite option.

Good points. It's also not my favorite.

But expending on my last point - which likely affects the change in this PR:

I think we have to ditch the `diagnostics` parameter for vim.lsp.buf.code_action, or change it to a `client -> diagnostics` function and create the code-action request params on a per-client basis. (There is more broken there, . The `make_range_params` call also doesn't set the encoding parameter, all of that needs to be client specific)

Assuming you've 2 clients connected to the buffer, and 2 diagnostics spanning the cursor position, 1 from each client then each client should only receive their own diagnostics.
And we should probably also fetch diagnostics for both pull and push namespaces.

Reply:
> Assuming you've 2 clients connected to the buffer, and 2 diagnostics spanning the cursor position, 1 from each client then each client should only receive their own diagnostics.

The user will still ultimately call `vim.lsp.buf.code_action()` which needs to aggregate code actions from all clients and display them together. Do you have any idea on how best to do that?

Given that multi client support is a non-trivial refactor I propose handling that in a separate PR. Any objection? This PR does not need to do anything to `get_line_diagnostics` except add the `@deprecated` tag (it already does not show up in the docs, so this is just enforcing its status on a documentation level).

Reply:
> I think we have to ditch the diagnostics parameter for vim.lsp.buf.code_action

@mfussenegger do you mean that we should hide it from the public API? Because the LSP request still needs diagnostics in the context parameter. You're saying this should all happen inside the implementation of `vim.lsp.buf.code_action`?

Reply:
Maybe to illustrate with some code - completely untested - but it should show what I mean with per client params and why I think it affects the changes here:



```diff
diff --git a/runtime/lua/vim/lsp/buf.lua b/runtime/lua/vim/lsp/buf.lua
index 9436fbbf5..f0e969f22 100644
--- a/runtime/lua/vim/lsp/buf.lua
+++ b/runtime/lua/vim/lsp/buf.lua
@@ -692,19 +692,10 @@ local function on_code_action_results(results, ctx, options)
       return title:gsub('\n', '\\n')
     end,
   }, on_user_choice)
 end
 
---- Requests code actions from all clients and calls the handler exactly once
---- with all aggregated results
-local function code_action_request(params, options)
-  local bufnr = api.nvim_get_current_buf()
-  vim.lsp.buf_request_all(bufnr, ms.textDocument_codeAction, params, function(results)
-    local ctx = { bufnr = bufnr, method = ms.textDocument_codeAction, params = params }
-    on_code_action_results(results, ctx, options)
-  end)
-end
 
 --- Selects a code action available at the current
 --- cursor position.
 ---
 ---@param options table|nil Optional table which holds the following optional fields:
@@ -738,33 +729,51 @@ function M.code_action(options)
   -- Detect old API call code_action(context) which should now be
   -- code_action({ context = context} )
   if options.diagnostics or options.only then
     options = { options = options }
   end
-  local context = options.context or {}
-  if not context.triggerKind then
-    context.triggerKind = vim.lsp.protocol.CodeActionTriggerKind.Invoked
+  local bufnr = api.nvim_get_current_buf()
+  local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_codeAction })
+  local remaining = #clients
+  if remaining == 0 then
+    return
   end
-  if not context.diagnostics then
-    local bufnr = api.nvim_get_current_buf()
-    context.diagnostics = vim.lsp.diagnostic.get_line_diagnostics(bufnr)
-  end
-  local params
   local mode = api.nvim_get_mode().mode
-  if options.range then
-    assert(type(options.range) == 'table', 'code_action range must be a table')
-    local start = assert(options.range.start, 'range must have a `start` property')
-    local end_ = assert(options.range['end'], 'range must have a `end` property')
-    params = util.make_given_range_params(start, end_)
-  elseif mode == 'v' or mode == 'V' then
-    local range = range_from_selection(0, mode)
-    params = util.make_given_range_params(range.start, range['end'])
-  else
-    params = util.make_range_params()
+  local win = api.nvim_get_current_win()
+  local results = {}
+  for _, client in ipairs(clients) do
+    local context = options.context and vim.deepcopy(options.context) or {}
+    if not context.triggerKind then
+      context.triggerKind = vim.lsp.protocol.CodeActionTriggerKind.Invoked
+    end
+    -- TODO:
+    -- context.diagnostics = vim.lsp.diagnostic._get(client.id)
+    local params
+    if options.range then
+      assert(type(options.range) == 'table', 'code_action range must be a table')
+      local start = assert(options.range.start, 'range must have a `start` property')
+      local end_ = assert(options.range['end'], 'range must have a `end` property')
+      params = util.make_given_range_params(start, end_, bufnr, client.offset_encoding)
+    elseif mode == 'v' or mode == 'V' then
+      local range = range_from_selection(0, mode)
+      params = util.make_given_range_params(range.start, range['end'], bufnr, client.offset_encoding)
+    else
+      params = util.make_range_params(win, client.offset_encoding)
+    end
+    params.context = context
+    client.request(ms.textDocument_codeAction, params, function(err, result)
+      results[client.id] = { error = err, result = result, params = params }
+      remaining = remaining - 1
+      if remaining == 0 then
+        local ctx = {
+          bufnr = bufnr,
+          method = ms.textDocument_codeAction,
+        }
+        on_code_action_results(results, ctx, options)
+      end
+    end, bufnr)
   end
-  params.context = context
-  code_action_request(params, options)
 end
 
 --- Executes an LSP server command.
 ---
 ---@param command_params table A valid `ExecuteCommandParams` object
 ```

Reply:
I'm sorry to pile more into this,  but I think there's one more issue: If the user is in visual mode or supplies a range,  I suspect it should use all diagnostics contained in that range, instead of those spanning the cursor position (?)

Reply:
Created https://github.com/neovim/neovim/pull/25873 to deal with the per-client params (excluding the per-client diagnostic lookup. That could then be done in this PR)